### PR TITLE
codeintel: Optimize UpdateDependencyNumReferences query

### PR DESCRIPTION
**Pull request by efritz**
This query should do the same thing but should choose a much better query plan. This should take some pressure off of Postgres as the previous query is inefficient enough to create some LSIF processing back pressure, which we generally don't encounter.

Two changes help here:

- We add a materialized cte around a select of a table to better hint at a more efficient index
- Use `UPDATE FROM` instead of a subselect, which kept scanning CTE results

I've been trying a few queries on the Cloud database and I've taken some samples down from minutes to sub-second with these changes.

### CODE REVIEW

File: enterprise/internal/codeintel/stores/dbstore/uploads.go [hunk 1]
```diff
@@ -768,46 +768,71 @@ func (s *Store) UpdateDependencyNumReferences(ctx context.Context, ids []int, de
...
+-- then join to references.
+--
+-- Instead, we materialize the CTE above to force Postgres to use the index
+-- on lsif_reference(dump_id). Then, the following query will use the index
+-- on lsif_packages(scheme, name, version, dump_id).
+--
+-- Note that each index operation in the latter case is targetted, i.e., has
+-- a specific b-tree target value. These perform better than full index scans
+-- which take time proportional to the entire database size and not the size
+-- of the result set, which is what we generally expect.
```
Comment (Strum355) [inline enterprise/internal/codeintel/stores/dbstore/uploads.go:810]:
```suggestion
-- If we do not materialize the CTE above, then we have a join over package
-- and reference tables, which Postgres likes to use a strange index for. By
-- default, Postgres wants to do a parallel (full) index scan over packages
-- then join to references.
--
-- Instead, we materialize the CTE above to force Postgres to use the index
-- on lsif_reference(dump_id). Then, the following query will use the index
-- on lsif_packages(scheme, name, version, dump_id).
--
-- Note that each index operation in the latter case is targeted, i.e., has
-- a specific b-tree target value. These perform better than full index scans
-- which take time proportional to the entire database size and not the size
-- of the result set, which is what we generally expect.
```

Comment (Strum355) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:810]: Im very not sure why the behaviour detailed is the way it is :thinking: is there a link or something to explain?

Comment (efritz) [reply enterprise/internal/codeintel/stores/dbstore/uploads.go:810]: > is that targeted is (american spelling) (target) while targetted is (british spelling).

lol wot

**Review by tjdevries — APPROVED** [review]
Change makes sense to me

**Review by Strum355 — APPROVED** [review]

**Pull request merged**